dask/bag/core.py

```diff
@@ -752,6 +759,103 @@ def write(data, filename):
         f.close()
 
 
+# private functions for from_s3
+def _get_s3_bucket(bucket_name, aws_access_key, aws_secret_key, connection,
+                   anon):
+    """Connect to s3 and return a bucket"""
+    import boto
+    if anon is True:
+        connection = boto.connect_s3(anon=anon)
+    elif connection is None:
+        connection = boto.connect_s3(aws_access_key, aws_secret_key)
+    return connection.get_bucket(bucket_name)
+
+
+# we need an unmemoized function to call in the main thread. And memoized
+# functions for the dask.
+_memoized_get_bucket = toolz.memoize(_get_s3_bucket)
+
+
+def _get_key(bucket_name, conn_args, key):
+    bucket = _memoized_get_bucket(bucket_name, *conn_args)
+    return bucket.get_key(key)
+
+
+def _parse_s3_URI(bucket_name, paths):
+    # FIXME
+    # DIRTY HACK: so that urlparse won't think '?' in the URI path is for a
+    # query. We replace '?' with 'QUESTIONMARK' so it stays in the path
+    # then after urlpars gives us the path replace 'QUESTIONMARK' with '?'.
+    # So this can be made to fail if a path has 'QUESTIONMARK' in it.
+    if '?' in bucket_name:
+        bucket_name = bucket_name.replace('?', 'QUESTIONMARK')
+    o = urlparse(bucket_name)
+    # if path is specified
+    if (paths == '*') and (o.path != '' and o.path != '/'):
+        paths = o.path[1:].replace('QUESTIONMARK', '?')
+    bucket_name = o.hostname
+    return bucket_name, paths
+
+
+def _from_s3(bucket_name, keys, conn_args):
+    """
+    Takes an s3 bucket name, list of bucket key, and connection info.
+    Returns a Bag.
+    """
+    name = next(load_names)
+    aak, ask, conn, anon = conn_args
+    # get the bucket
+    get_key = partial(_get_key, bucket_name, conn_args)
+    dsk = dict(((name, i), (list, (get_key, k))) for i, k in enumerate(keys))
+    return Bag(dsk, name, len(keys))
+
+
+def from_s3(bucket_name, paths='*', aws_access_key=None, aws_secret_key=None,
+            connection=None, anon=False):
+    """ Create a dask.bag by loading files from s3.
+
+    >>> b = from_s3(bucket, 'myfile1.txt')
+
+    or a pattern with '*' or '?'. There is one partition per file.
+
+    >>> b = from_s3(bucket, '*.json')
+    >>> b.npartitions
+    9
+
+    or a list of files.
+
+    >>> b = from_s3(bucket, ['alice.csv', 'bob.csv'])
+    >>> b.npartitions
+    2
+
+    """
+    conn_args = (aws_access_key, aws_secret_key, connection, anon)
+
+    if bucket_name.startswith('s3://'):
+        bucket_name, paths = _parse_s3_URI(bucket_name, paths)
+
+    # sing bucket key, or pattern to expand
+    if isinstance(paths, str):
+        if ('*' not in paths) and ('?' not in paths):
+            return _from_s3(bucket_name, [paths], conn_args)
+
+        # get the bucket
+        bucket = _get_s3_bucket(bucket_name, *conn_args)
+
+        # handle globs
+        keys = bucket.list()
```

Comment:
If possible I generally prefer comments at the end of a line.  My eyes sometimes have trouble walking over very vertical code.